src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php

```diff
@@ -0,0 +1,185 @@
+                throw new \InvalidArgumentException(sprintf('Invalid argument name "%s" used on the Argument annotation of %s::%s', $arg->getName(), $definition->getClass(), $reflectionMethod->getName()));
+            }
+            $key = $argumentIndexes[$arg->getName()];
+
+            $onInvalid = $arg->getOnInvalid();
+            $invalidBehavior = ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE;
+            if ('ignore' == $onInvalid) {
+                $invalidBehavior = ContainerInterface::IGNORE_ON_INVALID_REFERENCE;
+            } elseif ('null' == $onInvalid) {
+                $invalidBehavior = ContainerInterface::NULL_ON_INVALID_REFERENCE;
+            }
+
+            $type = $arg->getType();
+            // if "id" is set, default "type" to service
+            if (!$type && $arg->getId()) {
+                $type = 'service';
+            }
+
+            switch ($type) {
+                case 'service':
+                    $arguments[$key] = new Reference($arg->getId(), $invalidBehavior);
+                    break;
+                case 'expression':
+                    $arguments[$key] = new Expression($arg->getValue());
+                    break;
+                case 'closure-proxy':
+                    $arguments[$key] = new ClosureProxyArgument($arg->getId(), $arg->getMethod(), $invalidBehavior);
+                    break;
+                case 'collection':
+                    // todo
+                    break;
+                case 'iterator':
+                    // todo
+                    break;
+                case 'constant':
+                    $arguments[$key] = constant(trim($arg->getValue()));
+                    break;
+                default:
+                    $arguments[$key] = $arg->getValue();
+            }
+        }
+
+        // it's possible index 1 was set, then index 0, then 2, etc
+        // make sure that we re-order so they're injected as expected
+        ksort($arguments);
+
+        return $arguments;
+    }
+
+    /**
+     * @param \ReflectionMethod $method
+     *
+     * @return Annotations\Argument[]
+     */
+    private function getArgumentAnnotationsForMethod(\ReflectionMethod $method)
+    {
+        $annotations = $this->reader->getMethodAnnotations($method);
+        $argAnnotations = array();
+        foreach ($annotations as $annotation) {
+            if ($annotation instanceof Annotations\Argument) {
+                $argAnnotations[] = $annotation;
+            }
+        }
+
+        return $argAnnotations;
+    }
+
+    /**
+     * Returns arguments to a method, where the key is the *name*
+     * of the argument and the value is its index.
+     *
+     * @param \ReflectionMethod $method
+     *
+     * @return array
+     */
+    private function getMethodArguments(\ReflectionMethod $method)
+    {
+        $arguments = array();
+        $i = 0;
+        foreach ($method->getParameters() as $parameter) {
```

Comment:
Why not using the key instead of the `$i` variable?

```php
foreach($method->getParameters() as $i => $parameter) {
    $arguments[$parameter->getName()] = $i;
}
```

It should do the same thing: https://3v4l.org/Zpq4W